File: core/data/src/main/java/io/github/jeddchoi/data/service/SeatFinderRequest.kt
package io.github.jeddchoi.data.service

import io.github.jeddchoi.data.firebase.model.FirebaseSeatPosition
import kotlinx.serialization.EncodeDefault
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.Serializable

/**
 * Seat finder request
 *
 * @property requestType
 * @property seatPosition
 * @property endTime
 * @property durationInSeconds
 * @constructor Create empty Seat finder request
 */
@ExperimentalSerializationApi
@Serializable
data class SeatFinderRequest(
    val requestType: SeatFinderRequestType,
    @EncodeDefault val seatPosition: FirebaseSeatPosition? = null,
    @EncodeDefault val endTime: Long? = null,
    @EncodeDefault val durationInSeconds: Int? = null,
)

File: firebase/functions/src/https/request/SeatFinderRequest.ts
import {SeatFinderRequestType} from "./SeatFinderRequestType";
import {SeatPosition} from "../../firestore/seat/SeatPosition";

export class SeatFinderRequest {
    constructor(
        readonly requestType: SeatFinderRequestType,
        readonly seatPosition: SeatPosition | null,
        readonly durationInSeconds: number | null,
        private readonly _endTime: number | null
    ) {
    }

    static fromJson(json: {
        requestType: SeatFinderRequestType,
        seatPosition: SeatPosition | null,
        durationInSeconds: number | null,
        endTime: number | null
    }): SeatFinderRequest {
        return new SeatFinderRequest(
            json.requestType,
            json.seatPosition,
            json.durationInSeconds,
            json.endTime
        );
    }

    endTime(startTime: number) {
        if (this.durationInSeconds === null && this._endTime === null) {
            return null;
        } else if (this.durationInSeconds !== null) {
            return startTime + this.durationInSeconds * 1000;
        } else {
            return this._endTime;
        }
    }
}
